cudamapper/src/matcher_gpu.cuh

```diff
@@ -108,13 +148,196 @@ void compute_anchor_starting_indices(
     const device_buffer<std::int64_t>& found_target_indices_d,
     const device_buffer<std::uint32_t>& target_starting_index_of_each_representation_d);
 
+/// \brief Generates an array of anchors from matches of representations of the query and target index
+///
+/// See generate_anchors_dispatcher() for more details
+///
+/// \param anchors_d the array to be filled with anchors, the size of this array has to be equal to the last element of anchor_starting_indices
+/// \param compound_key_read_ids_d the array to be filled with compund keys, the size of this array has to be equal to the last element of anchor_starting_indices
+/// \param compound_key_positions_in_reads_d the array to be filled with compund keys, the size of this array has to be equal to the last element of anchor_starting_indices
+/// \param n_anchors the size of the anchors array
+/// \param anchor_starting_indices_d the array of starting indices of the set of anchors for each unique representation of the query index (representations with no match in target will have the same starting index as the last matching representation)
+/// \param query_starting_index_of_each_representation_d the starting index of a representation in query_read_ids and query_positions_in_read
+/// \param found_target_indices_d the found matches in the array of unique target representation for each unique representation of query index
+/// \param target_starting_index_of_each_representation_d the starting index of a representation in target_read_ids and target_positions_in_read
+/// \param n_query_representations the size of the query_starting_index_of_each_representation_d and found_target_indices_d arrays, ie. the number of unique representations in the query index
+/// \param query_read_ids the array of read ids of the (read id, position)-pairs in query index
+/// \param query_positions_in_read the array of positions of the (read id, position)-pairs in query index
+/// \param target_read_ids the array of read ids of the (read id, position)-pairs in target index
+/// \param target_positions_in_read the array of positions of the (read id, position)-pairs in target index
+/// \param smallest_query_read_id smallest read_id in query index
+/// \param smallest_target_read_id smallest read_id in target index
+/// \param number_of_target_reads number of read_ids in taget index
+/// \param max_basepairs_in_target_reads number of basepairs in longest read in target index
+/// \tparam ReadsKeyT type of compound_key_read_ids_d, has to be integral
+/// \tparam PositionsKeyT type of compound_key_positions_in_reads_d, has to be integral
+template <typename ReadsKeyT, typename PositionsKeyT>
+__global__ void generate_anchors_kernel(
```

Comment:
Do we really need/want to provide these templates in the `matcher_gpu.cuh` header file?

These functions look like implementation details of the `MatcherGPU` class to me (in contrast to generic algorithms/containers which a user/developer may use directly). If so, I would move the function templates to an anonymous namespace in the `matcher_gpu.cu` file, in order to reduce the amount of code that needs to get recompiled when we change something here.

Reply:
No, this makes sense. I'll move it all to `.cu`

Reply:
done